### PR TITLE
Compiler manager should use providers

### PR DESCRIPTION
**Pull request by cstamas**
To figure out presence of component,
and if construction fails, report some
sane(r) message tham obvious false
information: "compiler not present".

### CODE REVIEW

**Comment by cstamas** [comment]
Sisu in general is fully lazy: see https://eclipse.github.io/sisu.inject/apidocs/reference/org/eclipse/sisu/BeanEntry.html

The `Provider<C>` is JSR330 wrapper of that above, calling into `BeanEntry#getProvider()`. Presence of BeanEntry means "bean is registered" (i.e. via index). But, beans in Sisu are always lazily created.

Doing `Map<String, MyComponent>.get("akey")` builds upon these, and in case of construction failure it "pretends" that bean is not present, returns `null` as per Map contract. Note that all the map values are lazy, created ONLY when you do Map.get

File: plexus-compiler-manager/src/main/java/org/codehaus/plexus/compiler/manager/DefaultCompilerManager.java [hunk 1]
```diff
@@ -36,19 +37,27 @@
...
     // CompilerManager Implementation
     // ----------------------------------------------------------------------
 
     public Compiler getCompiler(String compilerId) throws NoSuchCompilerException {
-        Compiler compiler = compilers.get(compilerId);
+        // Provider<Class> is lazy
+        // presence of provider means component is present (but not yet constructed)
+        Provider<Compiler> compiler = compilers.get(compilerId);
 
+        // it exists: as provider is found
```
Comment (laeubi) [inline plexus-compiler-manager/src/main/java/org/codehaus/plexus/compiler/manager/DefaultCompilerManager.java:51]: Should that not mean that it is NOT there if it is null?

Comment (cstamas) [reply plexus-compiler-manager/src/main/java/org/codehaus/plexus/compiler/manager/DefaultCompilerManager.java:51]: yes, yes, bad worded comments but i hope the intent is clear :smile:

Comment (laeubi) [reply plexus-compiler-manager/src/main/java/org/codehaus/plexus/compiler/manager/DefaultCompilerManager.java:51]: Maybe remove or move down into the if case (code speaks for itself?)

**Comment by kriegaex** [comment]
Superseded by #361. Thank you so much, @cstamas.

**Pull request closed without merging**